# AOC D1

## Part one

Example data

3   4

4   3

2   5

1   3

3   9

3   3

Maybe the lists are only off by a small amount! To find out, pair up the numbers and measure how far apart they are. Pair up the smallest number in the left list with the smallest number in the right list, then the second-smallest left number with the second-smallest right number, and so on.

Within each pair, figure out how far apart the two numbers are; you'll need to add up all of those distances. For example, if you pair up a 3 from the left list with a 7 from the right list, the distance apart is 4; if you pair up a 9 with a 3, the distance apart is 6.

In the example list above, the pairs and distances would be as follows:

- The smallest number in the left list is 1, and the smallest number in the right list is 3. The distance between them is 2.
- The second-smallest number in the left list is 2, and the second-smallest number in the right list is another 3. The distance between them is 1.
The third-smallest number in both lists is 3, so the distance between them is 0.
The next numbers to pair up are 3 and 4, a distance of 1.
The fifth-smallest numbers in each list are 3 and 5, a distance of 2.
Finally, the largest number in the left list is 4, while the largest number in the right list is 9; these are a distance 5 apart.

To find the total distance between the left list and the right list, add up the distances between all of the pairs you found. In the example above, this is 2 + 1 + 0 + 1 + 2 + 5, a total distance of 11!

In [68]:
import pandas as pd
import numpy as np

In [78]:
# Read the data
df = pd.read_csv("data/AOC_1_1.csv", header=None, sep='  ')
df.head()

C:\Users\Klemen\AppData\Local\Temp\ipykernel_18492\2080562697.py:2: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  df = pd.read_csv("data/AOC_1_1.csv", header=None, sep='  ')


0      1
0  60236  87497
1  27507  18604
2  69810  73952
3  60448  56269
4  92139  61722

In [70]:
# Sort first and second column individual
df_sorted = df[[0]].sort_values(by=0).reset_index(drop=True)
df_sorted[1] = df[[1]].sort_values(by=1).reset_index(drop=True)

In [71]:
# calculate the sum of absolute diferences
np.abs(df_sorted[0] - df_sorted[1]).sum()

np.int64(1506483)

### Answer is 1506483

## Part 2

This time, you'll need to figure out exactly how often each number from the left list appears in the right list. Calculate a total similarity score by adding up each number in the left list after multiplying it by the number of times that number appears in the right list.

Here are the same example lists again:

3   4
4   3
2   5
1   3
3   9
3   3
For these example lists, here is the process of finding the similarity score:

The first number in the left list is 3. It appears in the right list three times, so the similarity score increases by 3 * 3 = 9.
The second number in the left list is 4. It appears in the right list once, so the similarity score increases by 4 * 1 = 4.
The third number in the left list is 2. It does not appear in the right list, so the similarity score does not increase (2 * 0 = 0).
The fourth number, 1, also does not appear in the right list.
The fifth number, 3, appears in the right list three times; the similarity score increases by 9.
The last number, 3, appears in the right list three times; the similarity score again increases by 9.
So, for these example lists, the similarity score at the end of this process is 31 (9 + 4 + 0 + 0 + 9 + 9).

Once again consider your left and right lists. What is their similarity score?

In [79]:
# Create column with number of repeated numbers
df = df.merge(pd.DataFrame(df[1].value_counts()).reset_index(), left_on=0, right_on=1, how='left')

In [80]:
df.head()

0    1_x  1_y  count
0  60236  87497  NaN    NaN
1  27507  18604  NaN    NaN
2  69810  73952  NaN    NaN
3  60448  56269  NaN    NaN
4  92139  61722  NaN    NaN

In [81]:
# Sum up the column multiplied with count
(df_temp["count"] * df_temp[0]).sum()

np.float64(23126924.0)